cob/cob.c

```diff
@@ -471,34 +471,109 @@ M0_UNUSED static char *cob_dom_id_make(char *buf, const struct m0_cob_domain_id
 }
 
 /**
-   Set up a new cob domain
-
-   Tables are identified by the domain id, which must be set before calling
-   this function.
+ * Load the COB trees.
  */
 int m0_cob_domain_init(struct m0_cob_domain *dom, struct m0_be_seg *seg)
 {
+	int                     rc;
+	struct m0_btree_op      b_op = {};
+
 	M0_ENTRY("dom=%p id=%"PRIx64"", dom, dom != NULL ? dom->cd_id.id : 0);
 
 	M0_PRE(dom != NULL);
 	M0_PRE(dom->cd_id.id != 0);
 
-	m0_be_btree_init(&dom->cd_object_index,   seg, &cob_oi_ops);
-	m0_be_btree_init(&dom->cd_namespace,	  seg, &cob_ns_ops);
-	m0_be_btree_init(&dom->cd_fileattr_basic, seg, &cob_fab_ops);
-	m0_be_btree_init(&dom->cd_fileattr_omg,   seg, &cob_omg_ops);
-	m0_be_btree_init(&dom->cd_fileattr_ea,    seg, &cob_ea_ops);
+	M0_ALLOC_PTR(dom->cd_object_index);
+	M0_ASSERT(dom->cd_object_index);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_open(dom->cd_oi_node,
+						    sizeof dom->cd_oi_node,
+						    dom->cd_object_index, seg,
+						    &b_op));
+	M0_ASSERT(rc == 0);
+
+	M0_ALLOC_PTR(dom->cd_namespace);
+	M0_ASSERT(dom->cd_namespace);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_open(&dom->cd_ns_node,
+						    sizeof dom->cd_ns_node,
+						    dom->cd_namespace, seg,
+						    &b_op));
+	M0_ASSERT(rc == 0);
+
+	M0_ALLOC_PTR(dom->cd_fileattr_basic);
+	M0_ASSERT(dom->cd_fileattr_basic);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_open(&dom->cd_fa_basic_node,
+						    sizeof dom->cd_fa_basic_node,
+						    dom->cd_fileattr_basic, seg,
+						    &b_op));
+	M0_ASSERT(rc == 0);
+
+	M0_ALLOC_PTR(dom->cd_fileattr_omg);
+	M0_ASSERT(dom->cd_fileattr_omg);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_open(&dom->cd_fa_omg_node,
+						    sizeof dom->cd_fa_omg_node,
+						    dom->cd_fileattr_omg, seg,
+						    &b_op));
+	M0_ASSERT(rc == 0);
+
+	M0_ALLOC_PTR(dom->cd_fileattr_ea);
+	M0_ASSERT(dom->cd_fileattr_ea);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_open(&dom->cd_fa_ea_node,
+						    sizeof dom->cd_fa_ea_node,
+						    dom->cd_fileattr_ea, seg,
+						    &b_op));
+	M0_ASSERT(rc == 0);
 
 	return M0_RC(0);
 }
 
 void m0_cob_domain_fini(struct m0_cob_domain *dom)
 {
-	m0_be_btree_fini(&dom->cd_fileattr_ea);
-	m0_be_btree_fini(&dom->cd_fileattr_omg);
-	m0_be_btree_fini(&dom->cd_fileattr_basic);
-	m0_be_btree_fini(&dom->cd_object_index);
-	m0_be_btree_fini(&dom->cd_namespace);
+	struct m0_btree_op b_op = {};
+
+	M0_ENTRY("dom=%p id=%"PRIx64"", dom, dom != NULL ? dom->cd_id.id : 0);
+
+	M0_PRE(dom != NULL);
+	M0_PRE(dom->cd_id.id != 0);
+
+	if (dom->cd_object_index != NULL) {
+		M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+					 m0_btree_close(dom->cd_object_index,
+							       &b_op));
```

Comment:
alignment. Please check the same for all m0_btree_close() functions at line no. 553, 560 etc.

Reply:
Resolving conversation on behalf of Radha.